specify return types where possible

cryy22 opened the pull request:
partially resolves #9

this PR introduces an update to the generator that attempts to specify function return types where possible, rather than always returning IntPtrs when the return value is a pointer. this is a disruptive change that alters lots of existing function signatures, and i believe it will blow up existing `Marshal.PtrToStructure` calls, so i'm not sure how we'd like to handle the rollout here.

this change does not attempt to marshal returned c-style arrays since it seems some additional attribute config is necessary (`SizeParamIndex` to inform the marshaller of the length of the array, i believe) -- i'll handle that in a future PR.

flibitijibibo commented:
The SDL2 policy was generally to keep IntPtrs and let people do the cast with `unsafe` for a couple reasons:

- Easy for apps that don't need the struct internals and want to avoid unsafe
- For apps that _can_ use unsafe, casting to/from IntPtr is thankfully not as bad as others since it's a value type
- Horrible early memories where Mono doesn't get the struct layout correct and causes random memory corruptions

The first two are definitely a matter of opinion but the third probably doesn't apply for Core. Since we're just _barely_ out in the wild this is probably our only chance to make a call on this... @thatcosmonaut thoughts?

cryy22 commented:
thanks for the context. it sounds like, given the mono horror stories, if we do choose to specify pointer return types, then we should confine that change to the core bindings. that would represent a significant divergence in the two libraries (i think they have identical type definitions and function signatures today). but it would also improve quality of life for the type of dev who is drawn to use SDL in a new project explicitly because of SDL3's release.

also this probably doesn't need to be said and it's not my call — but as with all code i am perfectly content with throwing it in the garbage

thatcosmonaut commented:
I agree, we can make this change for the Core bindings only. I don't think there's any expectation that you should be able to switch freely between the Legacy and Core bindings, the Core runtime is completely different anyway.

cryy22 commented:
updated this PR to only add specific reference return type data to the Core bindings, and leave the Legacy bindings alone.

i also added support for marshaling returned arrays, by specifying the name of the function parameter that is populated with the length of the array by SDL before returning.
